sw/host/opentitanlib/src/ownership/owner.rs

```diff
@@ -0,0 +1,594 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use serde::{Deserialize, Serialize};
+use serde_annotate::Annotate;
+use std::convert::TryFrom;
+use std::io::{Read, Write};
+
+use super::misc::{KeyMaterial, OwnershipKeyAlg, TlvHeader, TlvTag};
+use super::{OwnerApplicationKey, OwnerFlashConfig, OwnerFlashInfoConfig, OwnerRescueConfig};
+use crate::crypto::ecdsa::{EcdsaPrivateKey, EcdsaRawSignature};
+use crate::with_unknown;
+
+with_unknown! {
```

Comment:
What does `with_unkown!` do?

Reply:
With unknown is a utility macro for creating so-called Open Enums.

Rust enums are closed enums, meaning all un-enumerated values are not legal values.  C enums, OTOH, are open enums, meaning the enum _may_ take on values not enumerated in the definition.

The various `with_unknown` enums here are user-input, and thus might have values "unknown" to the enum definition (they _shouldn't_, but they _might_).

In order to accomplish this in Rust, the `with_unknown` macro rewrites `enum Foo { ... }` as `struct Foo(u32)` and then creates struct-level consts for all of the enumerators.  It also creates a number of trait implementations, including string conversions, Display, Debug, and serde Serialize/Deserialize.  These implementations make the enum useful in several contexts in opentitantool/lib (namely: CLI arguments and serialization to/from json).

Reply:
Thanks!